drivers/src/pci/device.rs

```diff
@@ -146,19 +149,295 @@ impl core::fmt::Display for PciDeviceInfo {
     }
 }
 
+// Returns a bit mask covering the specified number of bytes.
+fn byte_mask(bytes: usize) -> u32 {
+    ((1u64 << (bytes * 8)) - 1) as u32
+}
+
+// Returns the bytes in the specified range in `val`.
+fn select_bytes(val: u32, offset: usize, len: usize) -> u32 {
+    (val >> (offset * 8)) & byte_mask(len)
+}
+
+// Updates the bytes in `dest` in the specified range with `src`.
+fn update_bytes(dest: u32, offset: usize, len: usize, src: u32) -> u32 {
+    let mask = byte_mask(len) << (offset * 8);
+    dest & !mask | ((src << (offset * 8)) & mask)
+}
+
+// A builder for emulated MMIO config space read operations.
+//
+// In order to support partial accesses to word or dword registers, pushing a word or dword when
+// we're unaligned results in pushing only the bytes between the offset within the word/dword and
+// the end of the word/dword. For example, pushing a word register when `self.offset` is 1 results
+// in only the upper byte of the word being pushed into the result.
+struct MmioReadBuilder {
```

Comment:
nit: the MmioRead/WriteBuilder seems big enough to have its own file and maybe own commit in this series.

Reply:
Done.